gcp_variant_transforms/beam_io/bgzf_io.py

```diff
@@ -0,0 +1,176 @@
+# It beyonds the normal bin range (1-37449), and contains other meta data rather
+# than file offsets of GZIP blocks.
+_RESERVED_BIN_ID = 37450
+
+
+def split_bgzf(file_path):
+  # type: (str) -> Iterable[Tuple[str, Block]]
+  """Splits BGZF to multiple blocks.
+
+  It reads the `Block` (GZIP block offsets) from the corresponding index file.
+  Given (`file_path`, `Block`), one is able to decompress the data within
+  `Block`. It further:
+  - Removes invalid Blocks.
+  - Fills in the gap (by inserting a new Block) if there are gaps between two
+    adjacent Blocks to make sure it has full coverage of the variant data.
+  - Merges Blocks. It eliminates overlapping Blocks, and merges adjacent Blocks
+    if their combined size is smaller than `MAX_BLOCK_SIZE`.
+
+  Args:
+    file_path: The BGZF path.
+
+  Yields:
+    (file_path, Block).
+  """
+  blocks = _get_block_offsets(_get_tbi_file(file_path))
+  blocks = _remove_invalid_blocks(blocks)
+  blocks = _fill_in_gap(blocks)
+  blocks = _merge_blocks(blocks)
+
+  for block in blocks:
+    yield (file_path, block)
+
+
+def exists_tbi_file(file_path):
+  return filesystems.FileSystems.exists(_get_tbi_file(file_path))
+
+
+def _get_tbi_file(file_path):
+  return file_path + _INDEX_FILE_APPENDIX
+
+
+def _get_block_offsets(index_file):
+  # type: (str) -> List[Block]
+  """Returns block offsets by parsing the `index_file`.
+
+  The file format can be found in http://samtools.github.io/hts-specs/tabix.pdf.
+  """
+  read_buffer = _read(index_file)
+  offset = 4
+  n_ref, offset = _get_next_int(read_buffer, offset)
+  offset = 32
+  l_mn, offset = _get_next_int(read_buffer, offset)
+  offset += l_mn
+
+  chunk_virtual_file_offsets = []
+  for _ in range(n_ref):
+    n_bin, offset = _get_next_int(read_buffer, offset)
+    for bin_id in range(n_bin):
+      bin_id, offset = _get_next_unsigned_int(read_buffer, offset)
+      n_chunk, offset = _get_next_int(read_buffer, offset)
+      for _ in range(n_chunk):
+        start, offset = _get_next_unsigned_long_int(read_buffer, offset)
+        end, offset = _get_next_unsigned_long_int(read_buffer, offset)
+        if bin_id != _RESERVED_BIN_ID:
+          chunk_virtual_file_offsets.append((start, end))
+
+    n_intv, offset = _get_next_int(read_buffer, offset)
+    offset += 8 * n_intv
+
+  blocks = []
+  for chunk in chunk_virtual_file_offsets:
+    blocks.append(
+        Block(_get_block_offset(chunk[0]), _get_block_offset(chunk[1])))
+  return blocks
+
+
+def _read(index_file):
+  # type: (str) -> str
+  """Returns the contents of `index_file`."""
+  with filesystems.FileSystems.open(
+      index_file,
+      compression_type=filesystem.CompressionTypes.GZIP) as file_to_read:
+    lines = []
+    while True:
+      line = file_to_read.readline()
+      if not line:
+        break
+      lines.append(line)
+  return ''.join(lines)
+
+
+def _remove_invalid_blocks(blocks):
+  valid_blocks = []
+  for block in blocks:
+    if block.start != block.end:
```

Comment:
Why this particular case only?  What about end < start, etc?  I don't think, in general, this is likely, but if you are testing for one it seems reasonable to test for all?